crates/rate_limit/src/lib.rs

```diff
@@ -0,0 +1,279 @@
+use {
+    chrono::Duration,
+    deadpool_redis::{Pool, PoolError},
+    moka::future::Cache,
+    redis::{RedisError, Script},
+    std::{collections::HashMap, sync::Arc},
+};
+
+#[derive(Debug, thiserror::Error)]
+#[error("Rate limit exceeded. Try again at {reset}")]
+pub struct RateLimitExceeded {
+    reset: u64,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum InternalRateLimitError {
+    #[error("Redis pool error {0}")]
+    Pool(PoolError),
+
+    #[error("Redis error: {0}")]
+    Redis(RedisError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RateLimitError {
+    #[error(transparent)]
+    RateLimitExceeded(RateLimitExceeded),
+
+    #[error("Internal error: {0}")]
+    Internal(InternalRateLimitError),
+}
+
+/// Rate limit check using a token bucket algorithm for one key and in-memory
+/// cache for rate-limited keys. `mem_cache` TTL must be set to the same value
+/// as the refill interval.
+pub async fn token_bucket(
+    mem_cache: &Cache<String, u64>,
+    redis_write_pool: &Arc<Pool>,
+    key: String,
+    max_tokens: u32,
+    interval: Duration,
+    refill_rate: u32,
+    now_millis: i64,
```

Comment:
Would be better to pass a `DateTime` so we don't depend on implementers providing the right format.

Reply:
Totally, Thanks! Changed to `DateTime`.